Support calling fit and test scripts using "python -m" module syntax with DDP

nisheethlahoti opened the pull request:
## What does this PR do?
Fixes #8072
Fixes #3689 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? 
(not for typos, docs, test updates, or internal minor changes/refactorings)
[Note: Didn't update documentation because the existing failure was undocumented]

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

justusschock commented on pytorch_lightning/plugins/training_type/ddp.py:
```diff
@@ -155,19 +156,24 @@ def _call_children_scripts(self):
+            # use the same python interpreter and actually running
+            command = [sys.executable] + command
+        else:  # Script called as `python -m a.b.c`
+            command = [sys.executable, "-m", __main__.__spec__.name] + sys.argv[1:]
```
don't we need to find the hydra path here as well?

nisheethlahoti replied:
I haven't worked with Hydra, and don't know why it needs the original path, or how to resolve that requirement here. If you could suggest how to make hydra work with this case, I can add that to the pull request.

Alternatively, I can add `assert not _HYDRA_AVAILABLE` here and document that Hydra is not supported with "python -m" syntax. That would still be an improvement over the current state of things.

Please let me know how to move forward.

nisheethlahoti replied:
Never mind, I read about it, and the reason that we needed Hydra's to_absolute_path for the `python a/b.py` case is that hydra copies the file to a different location and runs the copied file instead of the main one, so `os.path.abspath` would have returned the wrong result.

But the module name `a.b` in this case isn't actually changed. And the new process will be able to find the module even using the original name because we're calling the subprocess with `cwd=get_original_cwd()`.

Just to be sure, I installed Hydra and tested with it, it all works.

carmocca commented on tests/accelerators/test_ddp.py:
```diff
@@ -45,9 +44,18 @@ def test_multi_gpu_model_ddp_fit_only(tmpdir):
 
 
 @RunIf(min_gpus=2)
-def test_multi_gpu_model_ddp_test_only(tmpdir):
+def test_multi_gpu_model_ddp_fit_only_asfile(tmpdir):
+    _test_multi_gpu_model_ddp_fit_only(tmpdir, as_module=False)
+
+
+@RunIf(min_gpus=2)
+def test_multi_gpu_model_ddp_fit_only_asmodule(tmpdir):
+    _test_multi_gpu_model_ddp_fit_only(tmpdir, as_module=True)
```
Can be parametrized. Same for the other tests added

```suggestion
@pytest.mark.parametrize('as_module', (False, True), marks=RunIf(min_gpus=2))
def test_multi_gpu_model_ddp_fit_only_asmodule(tmpdir, as_module):
    _test_multi_gpu_model_ddp_fit_only(tmpdir, as_module=as_module)
```

carmocca replied:
Can be parametrized. Same for the other tests added

```suggestion
@pytest.mark.parametrize('as_module', (False, True))
def test_multi_gpu_model_ddp_fit_only_asmodule(tmpdir, as_module):
    _test_multi_gpu_model_ddp_fit_only(tmpdir, as_module=as_module)
```

nisheethlahoti commented:
The IPU build is failing at some point completely unrelated to the PR: at "Reset IPU devices", which seems to be an infrastructure issue. Can it be run again or something?

carmocca commented:
> Can it be run again or something?

Done. It's resource-limited so might fail again. You can ping any of us in Slack if you need to re-run

nisheethlahoti commented:
Is there anything more that needs to be done to merge this PR? (auto-merge was automatically disabled when I added the remaining commits).

I'm not sure what to do about codecov, since I've already added the new variants in all the relevant tests.

awaelchli commented:
Yes, I need to test it. Let me have a look before merge :)

awaelchli commented:
Seems to work fine. I tested
```python
python -m "pl_examples.basic_examples.simple_image_classifier" --trainer.gpus 2 --trainer.accelerator ddp
```
However, note this also works on master. The changes here however seem to address the issue of relative imports. 
If we change the line
```python
from pl_examples.basic_examples.mnist_datamodule import MNISTDataModule
```
to a relative one
```python
from .mnist_datamodule import MNISTDataModule
```
Then on master we hang, and this branch works as expected as best as I can tell.

awaelchli commented on pytorch_lightning/plugins/training_type/ddp.py:
```diff
@@ -155,19 +156,24 @@ def _call_children_scripts(self):
-        command = [sys.executable] + command
+        # Check if the current calling command looked like `python a/b/c.py` or `python -m a.b.c`
+        # See https://docs.python.org/3/reference/import.html#main-spec
+        if __main__.__spec__ is None:  # Script called as `python a/b/c.py`
```
can codecov follow this code path? we can put a `# pragma: no-cover`

nisheethlahoti replied:
This path should be tested in the 3 modified tests with `as_module=False`. I don't know about the internals of how codecov works, so if it recognizes that this is in fact tested.

Should I just add that pragma?

awaelchli replied:
I think it is because we launch our tests with `coverage run blabla`, but these tests you saw then calls a regular python script from there, `python train.py ` etc so coverge doesn't run from there and we will run the code under
`if __main__.__spec__ is None:` but coverage is not captured. So I suggest:

`if __main__.__spec__ is None:  # pragma: no-cover`

awaelchli commented on pytorch_lightning/plugins/training_type/ddp.py:
```diff
@@ -155,19 +156,24 @@ def _call_children_scripts(self):
-        command = [sys.executable] + command
+        # Check if the current calling command looked like `python a/b/c.py` or `python -m a.b.c`
+        # See https://docs.python.org/3/reference/import.html#main-spec
+        if __main__.__spec__ is None:  # pragma: no-cover . Script called as `python a/b/c.py`
```
```suggestion
        if __main__.__spec__ is None:  # pragma: no-cover
            # script called as `python a/b/c.py`
```

awaelchli replied:
@nisheethlahoti the comment was good, I just meant to say to move it below the if, not to interfere with the pragma

nisheethlahoti replied:
The way codecov works, it skips any line that has "pragma: no-cover" as a substring. So the existing comment doesn't interfere with the pragma either. Anyway, moving.

awaelchli commented:
Great addition!
